File: 2.Settembre/41.FileStorage-20Settembre.js
//todo Dati una dimensione di file, un'unità di misura per la dimensione del file e la capacità del disco rigido in gigabyte (GB), restituisci il numero di file che il disco rigido può memorizzare utilizzando i seguenti vincoli:

//! L'unità di misura per la dimensione del file può essere byte ("B"), kilobyte ("KB") o megabyte ("MB").
//! Restituisci il numero di file interi che l'unità può contenere.
//! Utilizza le seguenti conversioni:
//! Unità Equivalente
//! 1 B 1 B
//! 1 KB 1000 B
//! 1 MB 1000 KB
//! 1 GB 1000 MB
//! Ad esempio, dati 500, "KB" e 1 come argomenti, determina quanti file da 500 KB possono essere contenuti in un disco rigido da 1 GB.

//? Tempo di esecuzione:
//* 10 minuti 

//?Codice scritto senza utilizzo di ChatGpt

function numberOfFiles(fileSize, fileUnit, driveSizeGb) {

  let b = 1
  let kb = b * 1000
  let mb = kb * 1000
  let gb = mb * 1000

  if(fileUnit==="B") fileUnit = b
  if(fileUnit==="KB") fileUnit = kb
  if(fileUnit==="MB") fileUnit = mb
  if(fileUnit==="GB") fileUnit = gb

  return  Math.floor((driveSizeGb * gb) / (fileSize * fileUnit));
}

console.log(numberOfFiles(4096, "B", 1.5))

//?Codice scritto da ChatGpt

//* ✅ Pro

//* Logica corretta

//* Conversioni chiare e coerenti (base 1000)

//* Gestisce bene le unità diverse

//* Uso corretto di Math.floor

//! ❌ Contro

//! Riusi fileUnit per due scopi (stringa → numero)

//! Troppe variabili quando basta una mappa

//! Nessuna validazione su unità non valide

//! Un po’ verboso

//? Versione Chat Gpt 

function numberOfFiles(fileSize, fileUnit, driveSizeGb) {
  const units = {
    B: 1,
    KB: 1000,
    MB: 1000 ** 2,
    GB: 1000 ** 3
  };

  return Math.floor((driveSizeGb * units.GB) / (fileSize * units[fileUnit]));
}


//todo Considerazioni: 

//! N.B.

File: Python/2.Settembre/41.File_storage-20Settembre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# //Prompt ChatGpt: questo è il modello. Invierò consegna esercizio e codice, e vorrei feedback strutturato

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Dati la dimensione di un file, l'unità di misura della dimensione del file
# e la capacità del disco rigido in gigabyte (GB), restituire il numero di file
# che il disco rigido può memorizzare.
#
# Vincoli:
# - unità: "B", "KB", "MB"
# - 1 KB = 1000 B
# - 1 MB = 1000 KB
# - 1 GB = 1000 MB
# - risultato: numero intero di file memorizzabili

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

def number_of_files(file_size, file_unit, drive_size_gb):
    b = 1;
    kb = b * 1000;
    mb = kb * 1000;
    gb = mb * 1000;

    match file_unit:
        case "B":
            return (drive_size_gb * gb) // (file_size * b)
        case "KB":
            return (drive_size_gb * gb) // (file_size * kb)
        case "MB":
            return (drive_size_gb * gb) // (file_size * mb)
        case "GB":
            return (drive_size_gb * gb) // (file_size * gb)

    return file_size

print(number_of_files(4096, "B", 1.5))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def number_of_files(file_size, file_unit, drive_size_gb):
    units = {
        "B": 1,
        "KB": 1000,
        "MB": 1000**2,
        "GB": 1000**3
    }

    file_size_bytes = file_size * units[file_unit]
    drive_size_bytes = drive_size_gb * units["GB"]

    return drive_size_bytes // file_size_bytes

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Logica corretta e funzionante.
# - Uso di match/case moderno e leggibile.
# - Gestione chiara delle unità di misura.
# - Rispetta correttamente le conversioni richieste.
# - Restituisce il numero intero di file memorizzabili.

# //! Contro
# - Variabili b, kb, mb, gb ridondanti: si possono evitare.
# - La struttura è più lunga del necessario.
# - Il caso "GB" nel match è matematicamente inutile
#   (si lavora già in byte nel disco).
# - `return file_size` finale non ha senso logico (fallback errato).
# - Possibile rischio di errore se file_unit non è valido.

# =========================================================
# //* Consigli
# =========================================================

# - Soluzione corretta ma migliorabile in pulizia.
# - Il punto chiave è standardizzare tutto in byte:
#       file_size_bytes
#       drive_size_bytes
# - Evita variabili di conversione “manuali” ripetute.
# - Il match/case va benissimo, ma puoi semplificare molto la logica.
# - Attenzione al fallback finale: meglio sollevare errore o gestire None.
# - In un colloquio tecnico questa soluzione sarebbe considerata corretta
#   ma un po’ verbosa.

# Valutazione: 8.5/10
#
# Corretta e robusta, ma migliorabile in compattezza e pulizia.